packages/multichain/src/scope/filter.ts

```diff
@@ -0,0 +1,115 @@
+import { type Hex } from '@metamask/utils';
+
+import { assertIsInternalScopeString, assertScopeSupported } from './assert';
+import { isSupportedMethod, isSupportedNotification } from './supported';
+import type {
+  InternalScopeString,
+  NormalizedScopeObject,
+  NormalizedScopesObject,
+} from './types';
+
+/**
+ * Groups a NormalizedScopesObject into two separate
+ * NormalizedScopesObject with supported scopes in one
+ * and unsupported scopes in the other.
+ * @param scopes - The NormalizedScopesObject to group.
+ * @param hooks - The hooks.
+ * @param hooks.isChainIdSupported - A helper that returns true if an eth chainId is currently supported by the wallet.
+ * @returns an object with two NormalizedScopesObjects separated by support.
+ */
+export const bucketScopesBySupport = (
+  scopes: NormalizedScopesObject,
+  {
+    isChainIdSupported,
+  }: {
+    isChainIdSupported: (chainId: Hex) => boolean;
+  },
+) => {
+  const supportedScopes: NormalizedScopesObject = {};
+  const unsupportedScopes: NormalizedScopesObject = {};
+
+  for (const [scopeString, scopeObject] of Object.entries(scopes)) {
+    assertIsInternalScopeString(scopeString);
+    try {
+      assertScopeSupported(scopeString, scopeObject, {
+        isChainIdSupported,
+      });
+      supportedScopes[scopeString] = scopeObject;
+    } catch (err) {
+      unsupportedScopes[scopeString] = scopeObject;
+    }
+  }
+
+  return { supportedScopes, unsupportedScopes };
+};
+
+/**
+ * Returns a NormalizedScopesObject with only
+ * scopes that are supported entirely as is.
+ * @param scopes - The NormalizedScopesObject to filter.
+ * @param hooks - The hooks.
+ * @param hooks.isChainIdSupported - A helper that returns true if an eth chainId is currently supported by the wallet.
+ * @returns a NormalizedScopesObject with only scopes that are currently supported.
+ */
+export const filterScopesSupported = (
+  scopes: NormalizedScopesObject,
+  {
+    isChainIdSupported,
+  }: {
+    isChainIdSupported: (chainId: Hex) => boolean;
+  },
+) => {
+  const { supportedScopes } = bucketScopesBySupport(scopes, {
+    isChainIdSupported,
+  });
+
+  return supportedScopes;
+};
+
+/**
+ * Returns a NormalizedScopeObject with
+ * unsupported methods and notifications removed.
+ * @param scopeString - The InternalScopeString for the scopeObject.
+ * @param scopeObject - The NormalizedScopeObject to filter.
+ * @returns a NormalizedScopeObject with only methods and notifications that are currently supported.
+ */
+const filterScopeObjectSupported = (
```

Comment:
If a scope object has supported methods and notifications can we say that the whole thing is "supported"? Maybe a bit of a stretch but it could allow us to use a less awkward name:

``` suggestion
const getSupportedScopeObject = (
```

Reply:
oof. It's a stretch, but let's run with it

Reply:
https://github.com/MetaMask/core/pull/4813/commits/dc0283ef7acf14b58c39fb3ec1a1c584d96012bb